adding Day1 support documentation (#648)
rh-ecosystem-edge/kernel-module-management: docs/mkdocs/documentation/day1_limited_option.md

# Day1 kernel module loading support

> KMM is day2 operator, meaning: kernel modules are loaded only after the complete initialization of a linux (RCHOS) server.

Reviewer: ```suggestion
KMM is a Day 2 operator: kernel modules are loaded only after the complete initialization of a Linux (RCHOS) server.
```

> Sometimes, kernel module needs to be loaded at the nuch earlier stage, since it is provides functionality that is needed at
> much eralier stage. KMM day1 functionality allows customer to load kernel module at much eralier stage, using MCO functionality
> of the OCP cluster.

Reviewer: ```suggestion
Sometimes though, kernel module needs to be loaded at a much earlier stage.
KMM's Day 1 functionality allows customer to load kernel module on boot, using the [Machine Config Operator (MCO)](https://docs.openshift.com/container-platform/4.13/post_installation_configuration/machine-configuration-tasks.html).
```

## Supported use-cases

> Day1 functionality supports only a limited number of use cases. Its main use case is to allow loading OOT kernel module
> prior to NetworkManager service initialization. It does not support loading kernel module at initramfs stage. The following 
> are the conditions needed by Day1 functionality:
> 1. kernel module is currently not loaded in the kernel
> 2. the in-tree kernel module is loaded into the kernel, but can be unloaded and replaced by the OOT kernel module.
>    It means that the in-tree module is not referenced by any other kernel modules
> 3. in order for day1 functionlity to work, the node must have an functional network interface, meaning: an in-tree kernel driver for that imterface.
>    The OOT kernel module can be network driver that will replace a functional network driver.

Reviewer: ```suggestion
KMM's Day 1 functionality supports only a limited number of use cases. Its main use case is to allow loading out-of-tree kernel modules
prior to NetworkManager service initialization. It does not support loading kernel module at initramfs stage.
The following are the conditions needed by Day1 functionality:
1. kernel module is currently not loaded in the kernel
2. the in-tree kernel module is loaded into the kernel, but can be unloaded and replaced by the OOT kernel module.
   This means that the in-tree module is not referenced by any other kernel modules
3. in order for the Day 1 functionlity to work, the node must have an functional network interface, meaning: an in-tree kernel driver for that interface.
   The OOT kernel module can be a network driver that will replace the functional network driver.
```

## Day1 OOT kernel module loading flow

> The loading of OOT kernel module utilises support of MCO.The following is the sequence of the flow:
> 1. Customer applies an MCO yaml to the existing running cluster. In order to identify the necessary nodes that need to be
>    update, customer must identify/create an appropriate MachineConfigPool
> 2. MCO applies reboots node by node. On any rebooted node, 2 new systemd services are deployed: pull service and load service
> 3. Load service is configured to run prior to NetworkConfiguration service. It tries to pull a predefined kernel module image
>    abd then, using that image, to unload an in-tree module and load an OOT kernel module. 
> 4. Pull service is configured to run after NetworkManager service. It checks if the pre-configured kernel module image is located
>    on the node file-system. If it is - then the service exists normally, and server continues with linux initialization process.
>    If not - it pull the image onto the node, and reboots the node afterwards

Reviewer: ```suggestion
The loading of OOT kernel module leverages MCO. The flow sequence is as follows:
1. Apply an MCO YAML manifest to the existing running cluster. In order to identify the necessary nodes that need to be
   updated, you must create an appropriate `MachineConfigPool` resource.
2. MCO applies reboots node by node. On any rebooted node, 2 new systemd services are deployed: pull service and load service
3. Load service is configured to run prior to NetworkConfiguration service. It tries to pull a predefined kernel module image
   and then, using that image, to unload an in-tree module and load an OOT kernel module. 
4. Pull service is configured to run after NetworkManager service. It checks if the pre-configured kernel module image is located
   on the node's filesystem. If it is - then the service exists normally, and server continues with the boot process.
   If not - it pulls the image onto the node, and reboots the node afterwards.
```

> ## Kernel Module Image
> Day1 functionality uses the same DTK based image, as the one used by usual KMM Day2 flow. OOT kenrel module should be located under /opt
> base directory

Reviewer: ```suggestion
## Kernel Module Image

The Day 1 functionality uses the same DTK based image that Day 2 KMM builds can leverage.
OOT kernel module should be located under `/opt/lib/modules/${kernelVersion}`.
```

> ## In-tree module replacement
> Day1 functionality always tries to replace the in-tree kernel module with OOT kernel module. In case in-tree is not loaded, it does not affect
> the flow, the service will just continue to load the OOT kernel module.  

Reviewer: ```suggestion
## In-tree module replacement

The Day1 functionality always tries to replace the in-tree kernel module with the OOT one.
If the in-tree kernel module is not loaded, the flow is not affected;, the service will proceed and load the OOT kernel module.
```

> ## MCO yaml creation
> KMM provides an API to create an MCO yaml need for day1 functionality:

Reviewer: ```suggestion
## MCO yaml creation

KMM provides an API to create an MCO YAML manifest for the Day 1 functionality:

```

```go
ProduceMachineConfig(machineConfigName, machineConfigPoolRef, kernelModuleImage, kernelModuleName string) (string, error)
```

> The returned output is a string representation of an MCO yaml to be applied. It is up to the customer to apply this yaml.
> The parameters are:
> 1. _machineConfigName_ - The name of the MCO yaml. It will be set as a name parameter of the metadata of MCO yaml
> 2. _machineConfigPoolRef_ - the machineConfigPool name that will be used in order to identify the targeted nodes
> 3. _kernelModuleImage_ - container image that includes the OOT kernel module.
> 4. _kernelModuleName_ - the name of the OOT kernel module. This parameter will be used both to unload the in-tree kernel module
>    ( if loaded into the kernel) and to load the OOT kernel module
> 
> The API is located under _pkg/mcproducer_ package of hte KMM source code. There is no need to KMM operator to be running in order for the customer
> to use Day1 functionality. Customer only need to import _pkg/mcproducer_ package into it's operator/utility code, call the API and to apply the produced
> MCO yaml onto the cluster

Reviewer: ```suggestion

The returned output is a string representation of the MCO YAML manifest to be applied.
It is up to the customer to apply this YAML.

The parameters are:

- `machineConfigName`: the name of the MCO YAML manifest. It will be set as the `name` parameter of the metadata of MCO YAML manifest.
- `machineConfigPoolRef`: the `MachineConfigPool` name that will be used in order to identify the targeted nodes
- `kernelModuleImage`: the name of the container image that includes the OOT kernel module.
- `kernelModuleName`: the name of the OOT kernel module. This parameter will be used both to unload the in-tree kernel module
   (if loaded into the kernel) and to load the OOT kernel module.

The API is located under `pkg/mcproducer` package of hte KMM source code.
There is no need to KMM operator to be running in order for the customer to use Day1 functionality.
Users only need to import the `pkg/mcproducer` package into their operator/utility code, call the API and to apply the produced
MCO YAML to the cluster.
```

Reviewer: Please also remove the trailing lines.

# Day1 kernel module loading support

KMM is a Day 2 operator: kernel modules are loaded only after the complete initialization of a Linux (RCHOS) server.
Sometimes though, kernel module needs to be loaded at a much earlier stage.
KMM's Day 1 functionality allows customer to load kernel module during Linux systemd initialization stage, using the [Machine Config Operator (MCO)](https://docs.openshift.com/container-platform/4.13/post_installation_configuration/machine-configuration-tasks.html).

## Supported use-cases

KMM's Day 1 functionality supports only a limited number of use cases. Its main use case is to allow loading out-of-tree kernel modules
prior to NetworkManager service initialization. It does not support loading kernel module at initramfs stage.

> The following are the conditions needed by Day1 functionality:
> 1. kernel module is currently not loaded in the kernel

Reviewer: Required for correct rendering:
```suggestion
The following are the conditions needed by Day 1 functionality:

1. kernel module is currently not loaded in the kernel
```

2. the in-tree kernel module is loaded into the kernel, but can be unloaded and replaced by the OOT kernel module.
   This means that the in-tree module is not referenced by any other kernel modules
3. in order for the Day 1 functionlity to work, the node must have an functional network interface, meaning: an in-tree kernel driver for that interface.
   The OOT kernel module can be a network driver that will replace the functional network driver.

## Day1 OOT kernel module loading flow

> The loading of OOT kernel module leverages MCO. The flow sequence is as follows:
> 1. Apply an MCO YAML manifest to the existing running cluster. In order to identify the necessary nodes that need to be

Reviewer: Required for correct rendering:
```suggestion
The loading of OOT kernel module leverages MCO. The flow sequence is as follows:

1. Apply an MCO YAML manifest to the existing running cluster. In order to identify the necessary nodes that need to be
```

   updated, you must create an appropriate `MachineConfigPool` resource.
2. MCO applies reboots node by node. On any rebooted node, 2 new systemd services are deployed: pull service and load service
3. Load service is configured to run prior to NetworkConfiguration service. It tries to pull a predefined kernel module image
   and then, using that image, to unload an in-tree module and load an OOT kernel module. 
4. Pull service is configured to run after NetworkManager service. It checks if the pre-configured kernel module image is located
   on the node's filesystem. If it is - then the service exists normally, and server continues with the boot process.
   If not - it pulls the image onto the node, and reboots the node afterwards.

## Kernel Module Image

The Day 1 functionality uses the same DTK based image that Day 2 KMM builds can leverage.
OOT kernel module should be located under `/opt/lib/modules/${kernelVersion}`.

## In-tree module replacement

> The Day1 functionality always tries to replace the in-tree kernel module with the OOT one.

Reviewer: ```suggestion
The Day 1 functionality always tries to replace the in-tree kernel module with the OOT one.
```
